docs/connect.html

```diff
@@ -543,6 +543,67 @@ <h6>ACL requirements</h6>
+        <li>Start your worker in a test environment with <code>HYBRID_FAIL</code>. If all plugins are compatible, startup will succeed. If at least one plugin is not compatible the worker will fail to start up, and all incompatible plugins will be listed in the exception.</li>
+    </ul>
+
+    <p>If the verification step succeeds, then your current set of installed plugins are compatible, and it should be safe to change the <code>plugin.discovery</code> configuration to <code>SERVICE_LOAD</code>. If you change the set of already-installed plugins, they may no longer be compatible, and you should repeat the above verification. If the verification fails, you must address the incompatible plugins before using the <code>SERVICE_LOAD</code> strategy.</p>
+
+    <h5><a id="connect_plugindiscovery_migrateartifact" href="#connect_plugindiscovery_migrateartifact">Operators: Artifact Migration</a></h5>
+
+    <p>As an operator of Connect, if you discover incompatible plugins, there are multiple ways to try to resolve the incompatibility. They are listed below from most to least preferable.</p>
+
+    <ol>
+        <li>Upgrade your incompatible plugins to the latest release version from your plugin provider.</li>
+        <li>Contact your plugin provider and request that they migrate the plugin to be compatible, following the <a href="#connect_plugindiscovery_migratesource">source migration instructions</a>, and then upgrade to the migrated version.</li>
+        <li>Migrate the plugin artifacts yourself using the included migration script.</li>
+    </ol>
+
+    <p>The migration script is located in <code>bin/connect-plugin-path.sh</code> and <code>bin\windows\connect-plugin-path.bat</code> of your Kafka installation. The script can migrate incompatible plugin artifacts already installed on your Connect worker's <code>plugin.path</code> by adding or modifying JAR or resource files. This is not suitable for environments using code-signing, as this may change the artifacts such that they will fail signature verification. View the built-in help with <code>--help</code>.</p>
+
+    <p>To perform a migration, first use the <code>list</code> subcommand to get an overview of the plugins available to the script. You must tell the script where to find plugins, which can be done with the repeatable <code>--worker-config</code>, <code>--plugin-path</code>, and <code>--plugin-location</code> arguments. The script will only migrate plugins present in the paths specified, so if you add plugins to your worker's classpath, then you will need to specify those plugins via one or more <code>--plugin-location</code> arguments.</p>
```

Comment:
This part seems a little unclear: "The script will only migrate plugins present in the paths specified, so if you add plugins to your worker's classpath, then you will need to specify those plugins via one or more <code>--plugin-location</code> arguments."

Maybe "Note that the script will ignore plugins present on the Connect worker's classpath unless they are contained in an explicitly-provided <code>--plugin-location</code>"?

Also, can you refresh my memory on why we've designed the script to ignore classpath plugins? It feels like this has made things more difficult, both for us (implementing the `connect-plugin-path` script) and users (adding a footgun for plugins installed directly on the classpath). IMO it's not too late to tweak this behavior if we decide it's better to take things in a different direction.

Reply:
> Maybe "Note that the script will ignore plugins present on the Connect worker's classpath unless they are contained in an explicitly-provided --plugin-location"?

:+1:

> Also, can you refresh my memory on why we've designed the script to ignore classpath plugins? It feels like this has made things more difficult, both for us (implementing the connect-plugin-path script) and users (adding a footgun for plugins installed directly on the classpath).

The reason that the script doesn't migrate classpath plugins is because those containing JARs are part of the live-executing java process while the migration is running. I don't think editing them while they are being used is well defined, and would open us up to more bugs overall.

And the script doesn't `list` them to have parity with `sync-manifests`. 

> IMO it's not too late to tweak this behavior if we decide it's better to take things in a different direction.

I committed to it in the KIP:

> This script will only list and migrate plugins which are on the plugin.path of a Connect worker, and are loaded in isolation. This script will not list or migrate plugins which are included on the classpath, and will assume that classpath plugins have manifests added through some other method.

I understand that we can change the KIP, but I don't think we should. I think we might also need to change the name of the script too :)

Reply:
Hmmm... I agree with the reservations about modifying JAR files in-flight. Do we have to actually write to those JAR files, though? Could we create/modify a service loader manifest on the file system instead of inside a JAR file?

I know this is a little inelegant but IMO it's worth considering since it would reduce the potential for footguns related to classpath plugins and would increase the general utility of the CLI tool.

Reply:
> Hmmm... I agree with the reservations about modifying JAR files in-flight. Do we have to actually write to those JAR files, though? Could we create/modify a service loader manifest on the file system instead of inside a JAR file?
> I know this is a little inelegant but IMO it's worth considering since it would reduce the potential for footguns related to classpath plugins and would increase the general utility of the CLI tool.

If we did this, we'd end up with the same problem one iteration later. Suppose we were able to find the kafka lib directory, and knew that we were running via `kafka-run-class.sh`, so the manifest shim jar would be picked up the next time the script is run. Because it's on the classpath, we wouldn't be able to mutate it safely, so we could only append manifests by adding new shim jars. If we added a bad shim jar, or a user removed a plugin and wanted sync-manifests to remove the shim, the script couldn't remove the shim itself without undermining it's running AppClassLoader.

We could add some processing in the shell/batch script before or after the JVM execution, but maintaining non-trivial migration logic in both platform-agnostic shell and windows batch sounds like an opportunity for a lot more footguns. I'm concerned enough about copy-pasting batch files without being able to test them myself, I don't think I could implement the necessary parts in batch in time for the release. And I certainly don't want to maintain 3 different copies of the same migration logic :)

Reply:
Thanks, wish we didn't have this discrepancy in behavior but I'm convinced that you've thoroughly investigated and there aren't any maintainable alternatives 👍